File: benchmarks/pandas/bench_indexers.py
"""Benchmark: FixedForwardWindowIndexer and VariableOffsetWindowIndexer on 100k rows"""
import json, time
import numpy as np
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

offsets = np.array([3 + (i % 5) for i in range(ROWS)], dtype=np.int32)

fixed_idx = pd.api.indexers.FixedForwardWindowIndexer(window_size=10)
var_idx = pd.api.indexers.VariableOffsetWindowIndexer(index=pd.date_range("2020", periods=ROWS, freq="D"), offset=pd.offsets.Day(3))

# Warm up with just fixed (VariableOffsetWindowIndexer needs DatetimeIndex in pandas)
for _ in range(WARMUP):
    fixed_idx.get_window_bounds(ROWS, min_periods=1, center=False, closed=None)

start = time.perf_counter()
for _ in range(ITERATIONS):
    fixed_idx.get_window_bounds(ROWS, min_periods=1, center=False, closed=None)
    var_idx.get_window_bounds(ROWS, min_periods=1, center=False, closed=None)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "indexers",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_indexers.ts
/**
 * Benchmark: FixedForwardWindowIndexer and VariableOffsetWindowIndexer on 100k rows
 */
import { FixedForwardWindowIndexer, VariableOffsetWindowIndexer } from "../../src/index.js";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const offsets = Int32Array.from({ length: ROWS }, (_, i) => 3 + (i % 5));

const fixedIdx = new FixedForwardWindowIndexer({ windowSize: 10 });
const varIdx = new VariableOffsetWindowIndexer({ indexArray: offsets });

// Warm up
for (let i = 0; i < WARMUP; i++) {
  fixedIdx.getWindowBounds(ROWS);
  varIdx.getWindowBounds(ROWS);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  fixedIdx.getWindowBounds(ROWS);
  varIdx.getWindowBounds(ROWS);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "indexers",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
